GPUSamplerDescriptor and GPUSampler creation spec

JusSn opened the pull request:
I'm sure I'm missing a lot! We've talked about just saving the entire descriptor in the `GPU` object before, but I don't think I saw another instance where we were doing that.


<!--
    This comment and the below content is programatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://pr-preview.s3.amazonaws.com/JusSn/gpuweb/pull/805.html" title="Last updated on Jun 8, 2020, 7:49 PM UTC (807b509)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/gpuweb/gpuweb/805/6d6e392...JusSn:807b509.html" title="Last updated on Jun 8, 2020, 7:49 PM UTC (807b509)">Diff</a>

Kangz commented on spec/index.bs:
```diff
@@ -1741,13 +1738,43 @@ enum GPUAddressMode {
+<dl dfn-type="enum-value" dfn-for=GPUAddressMode>
+    : <dfn>"clamp-to-edge"</dfn>
+    ::
+        Texture coordinates are clampled between 0.0 and 1.0, inclusive.
```
These will likely need a ton of expanding once we get to specifying how graphics operations work but look like a good initial description.

kvark replied:
Oh yes, this isn't exactly correct (see the comment on `GPUAddressMode`)

Kangz commented on spec/index.bs:
```diff
@@ -1761,6 +1788,80 @@ enum GPUCompareFunction {
+    Return true if and only if all of the following conditions apply:
+        - |device| is valid.
+        - |descriptor|.{{GPUSamplerDescriptor/lodMinClamp}} is greater than or equal to 0.
+        - |descriptor|.{{GPUSamplerDescriptor/lodMaxClamp}} is greater than or equal to
```
This should also disallow `NaNs`.

JusSn replied:
noob q, do IDL bindings for float not prevent this?

kainino0x replied:
> Infinity, -Infinity and NaN must not be used as the value of a float or double.

> The double type is a floating point numeric type that corresponds to the set of finite double-precision 64 bit IEEE 754 floating point numbers. [IEEE-754]

> The unrestricted double type is a floating point numeric type that corresponds to the set of all possible double-precision 64 bit IEEE 754 floating point numbers, finite, non-finite, and special "not a number" values (NaNs). [IEEE-754]

kainino0x commented on spec/index.bs:
```diff
@@ -1761,6 +1801,80 @@ enum GPUCompareFunction {
+    **Returns:** {{GPUSampler}}
+
+    1. Let |s| be a new {{GPUSampler}} object.
+    1. If |descriptor| is null or undefined:
+        1. Set |s|.{{GPUSampler/[[descriptor]]}} to a new {{GPUSamplerDescriptor}} object with specified default attributes.
+    1. Else:
+        1. Set |s|.{{GPUSampler/[[descriptor]]}} to |descriptor|.
```
I think we should adopt a convention where, here, `|descriptor|` is always defined and we don't need to write text to describe what happens. The rules for this are provided by WebIDL (since we have `= {}` above).
```suggestion
    1. Set |s|.{{GPUSampler/[[descriptor]]}} to |descriptor|.
```

JusSn replied:
Fixed!

kvark commented on spec/index.bs:
```diff
@@ -1715,23 +1727,21 @@ dictionary GPUSamplerDescriptor : GPUObjectDescriptorBase {
+- {{GPUSamplerDescriptor/magFilter}} specifies the sampling behavior for texture coordinates that do not
+    correspond to at least one texel.
+- {{GPUSamplerDescriptor/minFilter}} specifies the sampling behavior for texture coordinates that correspond
+    to multiple texels.
```
Yes. The texture coordinates alone do not correspond to one or more pixels. This is going into the whole derivatives story, so we can follow-up with clarifying this (by adding an Issue here).

JusSn replied:
Going to borrow Metal's "when the sample footprint is smaller/larger than a pixel"

kvark commented on spec/index.bs:
```diff
@@ -1715,23 +1727,21 @@ dictionary GPUSamplerDescriptor : GPUObjectDescriptorBase {
+- {{GPUSamplerDescriptor/mipmapFilter}} specifies behavior for sampling between two mipmap levels.
+- {{GPUSamplerDescriptor/lodMinClamp}} and {{GPUSamplerDescriptor/lodMaxClamp}} specify the minimum and
+    maximum levels of detail, respecitively, when sampling a texture.
+- If {{GPUSamplerDescriptor/compare}} is provided, the sampler will be a comparison sampler with the specified
```
would be good to explain in one on more sentences what "comparison sampler" actually means

kvark commented on spec/index.bs:
```diff
@@ -1715,23 +1727,21 @@ dictionary GPUSamplerDescriptor : GPUObjectDescriptorBase {
+    to multiple texels.
+- {{GPUSamplerDescriptor/mipmapFilter}} specifies behavior for sampling between two mipmap levels.
+- {{GPUSamplerDescriptor/lodMinClamp}} and {{GPUSamplerDescriptor/lodMaxClamp}} specify the minimum and
+    maximum levels of detail, respecitively, when sampling a texture.
```
this may need to be clarified, as "maximum levels of detail used by internally when sampling a texture in the shader". The "internal" bit is important, to avoid confusion with user-provided LOD level in some of the sampling functions.

kvark commented on spec/index.bs:
```diff
@@ -1715,23 +1727,21 @@ dictionary GPUSamplerDescriptor : GPUObjectDescriptorBase {
+- If {{GPUSamplerDescriptor/compare}} is provided, the sampler will be a comparison sampler with the specified
+    {{GPUCompareFunction}}.
+
+{{GPUAddressMode}} describes the behavior of the sampler if provided texture coordinates are beyond
```
This isn't true, strictly speaking. It affects the filtering on bounds as well, so the texture coordinate can be well within the texture, but the filtering footprint may not.

I think we'll need to specify what the "filtering footprint" is, to explain the filtering. It's fine to do in follow-ups, if we have an Issue here.

kvark commented on spec/index.bs:
```diff
@@ -1741,13 +1751,43 @@ enum GPUAddressMode {
+    : <dfn>"mirror-repeat"</dfn>
+    ::
+        Texture coordinates between -1.0 and 1.0, inclusive, are mirrored across the axis.
+        Outside -1.0 and 1.0, the texture coordinates wrap to the other side of the texture.
```
I don't think this is correct. What about, say, coordinate -2.5 ? We need to say something like: the texture is mirror expanded to -1.0 to 1.0 range, and that one is expanded. For example, this is how D3D12 describes it:
> Flip the texture at every (u,v) integer junction.
For u values between 0 and 1, for example, the texture is addressed normally; between 1 and 2, the texture is flipped (mirrored); between 2 and 3, the texture is normal again; and so on.

kvark commented on spec/index.bs:
```diff
@@ -1761,6 +1801,80 @@ enum GPUCompareFunction {
+<dl dfn-type="enum-value" dfn-for=GPUCompareFunction>
+    : <dfn>"never"</dfn>
+    ::
+        Comparison tests never pass.
```
what does it mean for the user that the comparison "passes", exactly?

JusSn replied:
Addressed just above this in the explanation for a "comparison sampler". WDYT?

JusSn commented:
For some reason, I'm not able to reply to the individual comments on this page right now, but I should have addressed concerns; let me know what y'all think.

kvark commented on spec/index.bs:
```diff
@@ -1762,12 +1764,12 @@ enum GPUAddressMode {
 
-{{GPUFilterMode}} describes the behavior of the sampler if the provided texture coordinates do not
-correspond to a single texel.
+{{GPUFilterMode}} describes the behavior of the sampler if the sample footprint is smaller or larger
```
that's not correct. The footprint can be exactly one texel (note: please use "texel" consistently), and we still need to filter it.
What we should be saying here is that "sample footprint does not exactly match a texel". This sounds similar to the previous statements, but there is a critical hidden difference: we are talking about the **shape** of the footprint here, where other places only talk about the **area** of the footprint. I think we need to change the other places to say "footprint area" when comparing to 1 everywhere, for clarity.

JusSn replied:
Hmm, I thought I replaced all mentions of "pixel" with "texel"; ok

kvark commented on spec/index.bs:
```diff
@@ -1786,7 +1788,9 @@ enum GPUFilterMode {
 
-{{GPUCompareFunction}} specifies the behavior of a comparison sampler.
+{{GPUCompareFunction}} specifies the behavior of a comparison sampler. If a comparison sampler is
+used in a shader, an input value is compared to the sampled texture value, and the result of this 
```
This happens to be much more nuanced that this. First of all, what is the result of comparison? The current text suggests that it's a boolean, when the truth it being a floating point value.
The other complication is that using any filtering with comparison means that the filtering will happen *on top* of the comparison results. I think we need to specify this carefully, possibly in another PR, but need to leave an issue here.

JusSn replied:
How about `0.0 for "fail", and 1.0 for "pass"`?

kvark replied:
yeah, that would be a good start!

kvark replied:
please leave an "Issue: describe how filtering interacts with comparison sampling"

JusSn commented on spec/index.bs:
```diff
@@ -1762,12 +1764,12 @@ enum GPUAddressMode {
-        Texture coordinates between -1.0 and 1.0, inclusive, are mirrored across the axis.
-        Outside -1.0 and 1.0, the texture coordinates wrap to the other side of the texture.
+        Texture coordinates wrap to the other side of the texture, but the texture is flipped
+        when the integer part of the coordinate is odd. 
```
The link you provided has contradicting answers about that; the top answer says parity applies to all integers. 0 is also an even number, but under some definitions it is not a natural number.

kvark replied:
ok, that's good

kainino0x replied:
I definitely think it's safe to assume even/odd are defined for integers.

JusSn commented on spec/index.bs:
```diff
@@ -1728,21 +1728,23 @@ dictionary GPUSamplerDescriptor : GPUObjectDescriptorBase {
-    correspond to at least one texel.
-- {{GPUSamplerDescriptor/minFilter}} specifies the sampling behavior for texture coordinates that correspond
-    to multiple texels.
+    and {{GPUSamplerDescriptor/addressModeW}} specify the address modes for the texture width,
```
Sorry, should be width, height, depth. fixing

JusSn commented on spec/index.bs:
```diff
@@ -1728,21 +1728,23 @@ dictionary GPUSamplerDescriptor : GPUObjectDescriptorBase {
+- {{GPUSamplerDescriptor/magFilter}} specifies the sampling behavior when the sample footprint
+    is smaller than one texel.
+- {{GPUSamplerDescriptor/minFilter}} specifies the sampling behavior when the sample footprint 
+    is larger than one texel.
```
Neither? Doesn't the sampling function return the value of that one texel?

JusSn replied:
Intuitively then, that's also minification? I can't find anything to back that up, though.

kvark replied:
@Kangz found the following piece:
> If λ is less than or equal to zero, the texture is said to be magnified, and the filter mode within a mip level is selected by the magFilter in the sampler. If λ is greater than zero, the texture is said to be minified, and the filter mode within a mip level is selected by the minFilter in the sampler.

So basically, it's magnification

kainino0x replied:
https://www.khronos.org/registry/vulkan/specs/1.2/html/vkspec.html#textures-texel-filtering

> If λ is less than or equal to zero, the texture is said to be magnified, and the filter mode within a mip level is selected by the magFilter in the sampler. If λ is greater than zero, the texture is said to be minified, and the filter mode within a mip level is selected by the minFilter in the sampler.

So in Vulkan apparently it's magnification.

kainino0x replied:
whoops, race condition :)

JusSn replied:
thanks!

JusSn replied:
Not sure if it's enough to spec here that it's magnification without offering an explanation of how LOD is calculated. left an issue.

JusSn commented:
uh oh. I think i messed up my branch

kainino0x commented on spec/index.bs:
```diff
@@ -1762,13 +1776,47 @@ enum GPUAddressMode {
+
+    : <dfn>"linear"</dfn>
+    ::
+        Select two texels in each dimension and return a linear interpolation between their values. 
```
nit: This makes it sound like, in *n* dimensions, 2*n* texels are interpolated, but isn't it 2<sup>*n*</sup>?
